Release notes v23.1.3 (#17230)
cockroachdb/docs: src/current/_includes/releases/v23.1/v23.1.3.md

- Fixed a bug that could cause goroutines to hang during [SCRAM authentication](../v23.1/security-reference/scram-authentication.html). [#104197][#104197]
- The `sys.cpu.combined.percent-normalized` metric now uses `GOMAXPROCS`, if lower than the number of CPU shares when calculating CPU utilization. [#104191][#104191]
- The warning message about a missing `--advertise-addr` flag is no longer displayed when the flag is specified on server start. [#104227][#104227]
- Fixed a rare bug where stale multi-column table [statistics](../v23.1/cost-based-optimizer.html) could cause table statistics forecasts to be inaccurate, leading to un-optimal query plans. [#104229][#104229]
- Fixed a bug in v23.1.0 where the `node_id` field would be omitted from logs. The `node_id` value has now been restored to the logs. [#103798][#103798]
- Fixed a bug where the collection of `KV bytes read` and `KV gRPC calls` execution statistics during [`EXPLAIN ANALYZE`](../v23.1/explain-analyze.html) could be incorrect (it would remain at zero) in some cases. The bug was introduced in the v23.1.0 release. [#104194][#104194]

> - Fixed an incorrect results bug where predicates on [computed columns](../v23.1/computed-columns.html) are derived when an ORed predicate on a column in the computed column expression is present. This bug only affects CockroachDB when the `optimizer_use_improved_computed_column_filters_derivation` [session setting](../v23.1/set-vars.html) is `true`. It currently defaults to `false` in releases v23.1.2 and v22.1.10. This bug does not affect releases prior to these versions. [#104276][#104276]

Reviewer: Vale has a good point -- can you write around using ORed as an adjective?

Author: Honestly, have not much of an idea how to re-write this. Looking at our release notes to date, we do use it as an adjective. It's not very common in our docs though...

I can only think of it adjectivally tbh "...when a predicate that is OR'ed..." I think maybe adding the apostrophe helps a little, but "OR'ed predicate" is more common in our docs and elsewhere on the web. 

I don't have any expertise in this area to do a quick rewrite! I can check in with the engineer on this, if it really sticks out to you. Let me know what you think.

Reviewer: Does it just mean "if the predicate contains OR"?

Reviewer: "if the predicate contains an OR condition on a computed column" perhaps?

Author: Cool, included

Reviewer: The more I look at this note -- it's not ideal to mention so many different versions in the same release note either. WDYT?

Author: Oh ok, I'm not sure I understand why it's not ideal? As I read over it initially I figured it was solid to be more precise here. But, if it's confusing in some way, I'm certainly open to changing.

> - Fixed an incorrect results bug where predicates on [computed columns](../v23.1/computed-columns.html) are derived when an ORed predicate on a column in the computed column expression is present. This bug only affects CockroachDB when the `optimizer_use_improved_computed_column_filters_derivation` [session setting](../v23.1/set-vars.html) is `true`. It currently defaults to `false` in releases v23.1.2 and v22.1.10. This bug does not affect releases prior to these versions. [#104276][#104276]

Reviewer: ```suggestion
- Fixed an incorrect results bug in v22.1.11 and 23.1.3 and above when the `optimizer_use_improved_computed_column_filters_derivation` [session setting](../v23.1/set-vars.html) is `true`. Predicates on [computed columns](../v23.1/computed-columns.html) are derived when an ORed predicate on a column in the computed column expression is present. [#104276][#104276]
```
WDYT about this edit? I'm still not sure that we should mention 22.1 in this note (if it will be fixed in a 22.1.x version) and I'm not 💯 this is still accurate to the intent. I still don't love ORed but also not sure if my suggestion for that bit is accurate either.

## v23.1.3

Release Date: June 13, 2023

{{site.data.alerts.callout_danger}}

> A change included in this patch release was discovered to [contain a bug](https://github.com/cockroachdb/cockroach/issues/104798) that can affect clusters upgrading from [v22.2.x](v22.2.html). This can result in any jobs that were running during the upgrade to hang or fail to run after the upgrade is finalized. Users upgrading from v22.2.x are advised to use [v23.1.2](#v23-1-2) to upgrade, or to set the [`cluster.preserve_downgrade_option`](../v23.1/upgrade-cockroach-version.html#step-3-decide-how-the-upgrade-will-be-finalized) cluster setting to prevent finalization of the upgrade until they can do so using v23.1.4.

Reviewer: ```suggestion
A [bug](https://github.com/cockroachdb/cockroach/issues/104798) was discovered in a change included in this patch  that can affect clusters upgrading from [v22.2.x](v22.2.html). In an affected cluster, jobs that were running during the upgrade could hang or fail to run after the upgrade is finalized. Users upgrading from v22.2.x are advised to use [v23.1.2](#v23-1-2) to upgrade, or to set the [`cluster.preserve_downgrade_option`](../v23.1/upgrade-cockroach-version.html#step-3-decide-how-the-upgrade-will-be-finalized) cluster setting to delay finalization of the upgrade until they can upgrade to v23.1.4.
```
WDYT? This might be a little more direct and simpler to read.

<h3 id="v23-1-3-security-updates">Security updates</h3>

- Added a new `server.client_cert_expiration_cache.capacity` [cluster setting](../v23.1/cluster-settings.html), which when set to a non-zero number will store the minimum time-until-expiration for the set of client certificates per user. This setting can be used to export client certificate expirations as a metric. For example, if set to zero, the `security.certificate.expiration.client` metric will have a value of zero. [#104165][#104165]

<h3 id="v23-1-2-{{-site.data.products.enterprise-}}-edition-changes">{{ site.data.products.enterprise }} edition changes</h3>

- Role-based audit logging is now an {{ site.data.products.enterprise }} feature. Only {{ site.data.products.enterprise }} users will be able to configure role-based audit logging using the `sql.log.user_audit` [cluster setting](../v23.1/cluster-settings.html). [#104453][#104453]

> - Introduced role-based auditing through the new `sql.log.user_audit` [cluster setting](../v23.1/cluster-settings.html). Using the cluster setting, a user is able to set an audit logging configuration using a table-like syntax. Each row in the configuration represents an _audit setting_ in the configuration. An audit setting is comprised of the following columns: `USER/ROLE` and `STATEMENT_FILTER`:

Author: This release note I have tried to summarize from the longer original. There aren't yet docs on this, so maybe we should include the full original draft (edited).

Reviewer: Makes sense.

    SET CLUSTER SETTING sql.log.user_audit = '
        test_role       ALL
        another_role    NONE
        ALL             NONE
    ';
    ~~~

    In this example, users with the username/role `test_role` will emit audit logs for all statements they issue. Users with the username/role `another_role` will not emit audit logs for any statements they issue. All remaining users will not emit audit logs for any statements they issue.

    To determine whether a statement should be emitted to the audit log, CockroachDB checks whether there is a corresponding audit setting for the statement (i.e., the statement filter must be `ALL`). At the **first** audit setting match (based on role/username), the check iteration will stop. Then the statement is added to a collection of audit events to be logged later. If the statement filter is `NONE`, the check stops. As a result, the order in which the audit settings in the configuration are specified is important, as only the **first matching audit setting** will apply. Note that this is an initial implementation and is subject to change and development. [#104179][#104179]

<h3 id="v23-1-3-sql-language-changes">SQL language changes</h3>

> - Fixed [`crdb_internal.reset_sql_stats()`](../v23.1/crdb-internal.html#statement-statistics) to clean up the `sql_activity` table that works as a cache for the statistics. [#104361][#104361]

Author: I had kept this because it was visible in docs i.e. crdb_internal.reset_sql_stats(), but re-reading the PR it does seem like this is not user facing

    - Ensure applications using the cluster have properly configured [connection pools](../v23.1/connection-pooling.html).
    - Add more vCPU or more nodes to the cluster.
    - Increase the password hashing concurrency using the `COCKROACH_MAX_PW_HASH_COMPUTE_CONCURRENCY` [environment variable](../v23.1/cockroach-commands.html#environment-variables). [#104388][#104388]

<h3 id="v23-1-3-db-console-changes">DB Console changes</h3>

- Added merge queue failure and merge queue processing time metrics to the [**Queue Processing Failures**](../v23.1/ui-queues-dashboard.html#queue-processing-failures) and [**Queue Processing Times**](../v23.1/ui-queues-dashboard.html#queue-processing-times) graphs respectively. [#104034][#104034]

> - Removed the **Circuit Breaker Tripped Events** graph from the [**Replication**](../v23.1/ui-replication-dashboard.html) dashboard. [#104038][#104038]

Author: I kind of felt like it did — I could see someone who used that graph wanting to know when it was released... maybe I'm wrong though!

- The [**Databases**](../v23.1/ui-databases-page.html) page now supports a large number of tables for a single database. If more than 40 tables are present in a database, the sort on the page will be disabled; however, it is still possible to filter by table name. [#103854][#103854]
- On the **Statement Details** page, renamed the "Idle Latency" metric to "Client Wait Time" and separated it into its own chart. [#103915][#103915]
- Added more search criteria options on the [**SQL Activity**](../v23.1/ui-overview.html#sql-activity) page:
    - For the **Top** dropdown: `1000`, `5000`, and `10000`.
    - For the **By** dropdown on the **Statements** tab: Last Execution Time, Max Latency, Max Memory, Min Latency, Network, P50 Latency, P90 Latency, Retries, and Rows Processed.
    - For the **By** dropdown on the **Transactions** tab: Max Memory, Network, Retries, and Rows Processed. [#104054][#104054]
- Added a new link to the **Range Status** page on the [**Replication**](../v23.1/ui-replication-dashboard.html) dashboard that opens the **Enqueue Ranges** page with the node ID already completed. [#104100][#104100]

> - The DB Console will no longer show `DROP INDEX` recommendations for unqiue indexes. [#104287][#104287]

Reviewer: ```suggestion
- The DB Console will no longer show `DROP INDEX` recommendations for unique indexes. [#104287][#104287]
```

Author: hmmm, weird a lot of these comments are outdated.... I have fixed these?

Reviewer: Yeah I think you sent a commit when I was reviewing your last one so we crossed the streams. I've already approved.

- Fixed an incorrect results bug where predicates on [computed columns](../v23.1/computed-columns.html) are derived when an ORed predicate on a column in the computed column expression is present. This bug only affects CockroachDB when the `optimizer_use_improved_computed_column_filters_derivation` [session setting](../v23.1/set-vars.html) is `true`. It currently defaults to `false` in releases v23.1.2 and v22.1.10. This bug does not affect releases prior to these versions. [#104276][#104276]
- The [`keyvisualizer` job](../v23.1/ui-key-visualizer.html) no longer panics if an error is encountered while cleaning up stale samples. Instead, if the job encounters an error, the job will try again later. [#104374][#104374]
- Fixed a bug where it was possible for a SQL row to be split across two ranges. When this occurred, SQL queries could return unexpected errors. The real keys are now inspected rather than just request keys to determine load-based split points. [#103876][#103876]

> - Fixed a bug which could cause a panic when a [`CREATE FUNCTION`](../v23.1/create-function.html) statement used the [`setval()` builtin function](../v23.1/functions-and-operators.html). [#104297][#104297]

Reviewer: ```suggestion
- Fixed a bug which could cause a panic when a [`CREATE FUNCTION`](../v23.1/create-function.html) statement used the [`setval()` built-in function](../v23.1/functions-and-operators.html). [#104297][#104297]
```

- Fixes a bug in which [`SHOW BACKUP`](../v23.1/show-backup.html) would fail to show a [locality-aware backup](../v23.1/take-and-restore-locality-aware-backups.html) that contained [incremental backups](../v23.1/take-full-and-incremental-backups.html). [#103761][#103761]

<h3 id="v23-1-3-performance-improvements">Performance improvements</h3>

- If [`transaction_rows_read_err`](../v23.1/transactions.html#limit-the-number-of-rows-written-or-read-in-a-transaction) is set to a non-zero value, CockroachDB now ensures that any single scan never reads more than `transaction_rows_read_err` + 1 rows. This prevents transactions that would error due to the `transaction_rows_read_err` setting from causing a large performance overhead due to large scans. [#104364][#104364]
- Improved the efficiency of the an internal job that during upgrades deleted descriptors of dropped functions. Previously, the job would check every single ID until the max descriptor ID, which was particularly inefficient when this was large. Now, the job only queries the upper bound ID of each batch. [#104590][#104590]

<h3 id="v23-1-3-contributors">Contributors</h3>

This release includes 111 merged PRs by 41 authors.

> </div>

Reviewer: Stray?
